### PR TITLE
[IMP] web_editor, website: control hue and saturation of grays

### PR DESCRIPTION
**Pull request by oussamamessaoudi**
task-2362091

### CODE REVIEW

File: addons/web/static/lib/bootstrap/scss/_variables.scss [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
 // Color system
 
+$base-gray-color-hue: 210 !default;
+$base-gray-color-saturation: 0 !default;
```
Comment (qsm-odoo) [inline addons/web/static/lib/bootstrap/scss/_variables.scss:9]: This is a lib file, you cannot modify it :grimacing: Especially since that feature has nothing to do with the web app, it is a website only feature. Handling colors is actually done in web_editor right now, this is where we set the gray palette, check for the `$o-gray-color-palette` variable. This is the map you should tweak, in our own files, not the bootstrap ones.

File: addons/web/static/lib/bootstrap/scss/_variables.scss [hunk 2]
```diff
@@ -34,6 +37,30 @@ $grays: map-merge(
   $grays
 );
 
+@function adjust-gray-color($gray) {
+  $base-color: change-color($gray, $hue: $base-gray-color-hue * 1deg);
+  @if $base-gray-color-saturation < 0 {
+    @return desaturate($base-color, $base-gray-color-saturation * -1%);
+  } @else {
+    @return saturate($base-color, $base-gray-color-saturation * 1%);
+  }
```
Comment (qsm-odoo) [inline addons/web/static/lib/bootstrap/scss/_variables.scss:46]: Can't you do that in one line with `adjust-color`?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -2090,7 +2090,7 @@ const userValueWidgetsRegistry = {
 const SnippetOptionWidget = Widget.extend({
     tagName: 'we-customizeblock-option',
     events: {
-        'click .o_we_collapse_toggler': '_onCollapseTogglerClick',
+        'click .o_we_collapse_toggler, .o_we_toggler': '_onCollapseTogglerClick',
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:2093]: Why not using the o_we_collapse_toggler instead of creating a new one?

File: addons/website/static/src/scss/website.editor.ui.scss [hunk 4]
```diff
@@ -60,3 +60,54 @@ $o-we-switch-inactive-color: #F7F7F7 !default;
...
+    &::-moz-range-thumb {
+        appearance: none !important;
+        height: 20px !important;
+        border: 1px solid $o-we-bg-dark !important;
+        box-shadow: inset 0 0 0 1px white !important;
+        background: transparent !important;
+        border-radius: 0 !important;
+        margin-top: -5% !important;
+    }
+}
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.editor.ui.scss:97]: I am pretty sure there is code to share the colorpalette slider here ;) Don't just copy/paste code, always try to reuse/share it.

File: addons/website/static/src/scss/website.editor.ui.scss [hunk 5]
```diff
@@ -60,3 +60,54 @@ $o-we-switch-inactive-color: #F7F7F7 !default;
...
+.o_we_toggler{
+    cursor: pointer;
+}
+.o_we_gray_preview {
+    span {
+        flex: 1;
+        margin: 0 !important;
+        height: 20px;
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.editor.ui.scss:105]: There are variable for this height

File: addons/website/views/snippets/snippets.xml [hunk 6]
```diff
@@ -1206,6 +1206,22 @@
             <we-colorpicker title="Warning" data-customize-website-color="" data-color-type="theme" data-color="warning"/>
             <we-colorpicker title="Error" data-customize-website-color="" data-color-type="theme" data-color="danger"/>
         </we-row>
+        <we-collapse>
+            <we-row string="Grays" class="o_we_gray_preview o_we_toggler">
+                <t t-foreach="9" t-as="i">
+                    <span t-att-title="'Gray '+ str((9-i)*100)"
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1212]: For this kind of stuff, use `t-attf-` instead of `t-att-`. It is easier to ready, and without it this won't be translated

File: addons/website/static/src/scss/primary_variables.scss [hunk 7]
```diff
@@ -305,6 +305,9 @@ $o-base-website-values-palette: (
     'footer-template': 'default',
     'footer-effect': null, // null / 'slideout_slide_hover' / 'slideout_shadow'
     'footer-scrolltop': false,
+
+    'base-gray-color-hue': 210,
+    'base-gray-color-saturation': 0
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/primary_variables.scss:310]: Let's name those `gray-hue` and `gray-extra-saturation`

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Commit message:
- wrong apps mentioned
- max 72 or 80 (depending who you ask) per line

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 3 above)
Comment (oussamamessaoudi) [reply addons/web_editor/static/src/js/editor/snippets.options.js:2093]: if i use `.o_we_collapse_toggler` i have to replace all css that contain, so i decided to create a new class.

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:2093]: Ok, then instead of this, just confine the scss to `we-toggler.o_we_collapse_toggler`

So, only one class: `.o_we_collapse_toggler` which has a JS handler attached to it
And the style you talked about is only applied to `we-toggler.o_we_collapse_toggler`

File: addons/website/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -797,7 +797,20 @@ options.registry.OptionsTab = options.Class.extend({
...
-
+    /**
+     * @see this.selectClass for parameters
+     */
+    async customizeWebsiteGrayParam(previewMode, widgetValue, params) {
+        if (!previewMode) {
+            await this._customizeWebsite(previewMode, widgetValue, params, 'variable');
+            return;
+        }
+        this.$el.find("span#grays").each((_, e) => {
```
Comment (oussamamessaoudi) [inline addons/website/static/src/js/editor/snippets.options.js:808]: @qsm-odoo
This part of the code is to update the preview of the gray palette.
I have a doubt about this any suggestions!

Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:808]: The UI update is to be done... in the `updateUI` method :wink: The option methods are to customize the snippet or the option's state: in this case, updating an internal variable for hue / saturation (`this.grayHue = ...`) (and maybe already compute the new grays). Then in updateUI, which is called automatically, you can check those internal variable and adapt the UI accordingly.

Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:808]: @oussamamessaoudi You can reply to comments on PR directly instead of linking me the PR in discord and asking me there. It makes it hard to follow Github discussions afterwards otherwise :wink:

Mmh indeed my bad, for preview mode you can indeed update the UI directly in that method then :+1:

File: addons/website/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -927,7 +940,21 @@ options.registry.OptionsTab = options.Class.extend({
...
-
+    /**
+     * @private
+     * @param {String} id
+     * @param {Object} settings: gray settings that will be adjusted to
+     * @returns {String} the new color of gray
+     */
+    _buildGray(id, settings) {
+        this['gray-hue'] = settings['gray-hue'] || this['gray-hue'];
+        this['gray-extra-saturation'] = settings['gray-extra-saturation'] || this['gray-extra-saturation'];
```
Comment (sdegueldre) [inline addons/website/static/src/js/editor/snippets.options.js]: Any particular reason these fields are named in kebab-case instead of camelCase?

File: addons/web_editor/static/src/js/common/utils.js [hunk 10]
```diff
@@ -233,6 +233,54 @@ function _getBgImageURL(el) {
...
+    }
+
+    s = s * 100;
+    s = Math.round(s);
+    l = l * 100;
+    l = Math.round(l);
+    h = Math.round(360 * h);
+
+    return {h, s, l};
+}
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/utils.js]: You can already do all of that using the methods in colorpicker.js

File: addons/web_editor/static/src/scss/secondary_variables.scss [hunk 11]
```diff
@@ -48,7 +48,18 @@ $o-theme-color-palette: map-merge($o-base-theme-color-palette, o-map-omit($o-the
...
+@function adjust-gray-color($gray) {
+    @return adjust-color(change-color($gray, $hue: $gray-hue * 1deg), $saturation: $gray-extra-saturation * 1%);
+}
+$o-adjusted-gray-color-palette: ();
+@each $key, $color in $o-base-gray-color-palette {
+    $o-adjusted-gray-color-palette: map-merge($o-adjusted-gray-color-palette,
+            (base-#{$key}: $color, $key: adjust-gray-color($color))
+    );
+}
+$o-gray-color-palette: map-merge($o-adjusted-gray-color-palette, o-map-omit($o-gray-color-palette));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/secondary_variables.scss]: The values in the grays palette are use to generate classes. If you had new keys "base-XXX" inside... there will be useless classes which are generated.

Also I think we will get rid of the multiple gray color palette at some point, but meanwhile there is not point in taking the grays from a palette and mixing them with base grays whose hue and saturation are tweaked. Compute the palette just like before:
```scss
$o-gray-color-palette: map-merge($o-base-gray-color-palette, o-map-omit($o-gray-color-palette));
```
And then adjust each gray inside that one.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 12]
```diff
@@ -927,7 +940,21 @@ options.registry.OptionsTab = options.Class.extend({
     //--------------------------------------------------------------------------
     // Private
     //--------------------------------------------------------------------------
-
+    /**
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:953]: Add a blank line before that

File: addons/website/views/snippets/snippets.xml [hunk 13]
```diff
@@ -1206,6 +1206,24 @@
             <we-colorpicker title="Warning" data-customize-website-color="" data-color-type="theme" data-color="warning"/>
             <we-colorpicker title="Error" data-customize-website-color="" data-color-type="theme" data-color="danger"/>
         </we-row>
+        <we-collapse>
+            <we-row string="Grays" class="o_we_gray_preview o_we_collapse_toggler">
+                <t t-foreach="9" t-as="i">
+                    <span t-attf-title="Gray #{str((9-i)*100)}"
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:1212]: Put `str((9-i)*100)` in a variable. You repeat it 3 times per loop

File: addons/website/static/src/scss/website.editor.ui.scss [hunk 14]
```diff
@@ -60,3 +60,62 @@ $o-we-switch-inactive-color: #F7F7F7 !default;
...
+        flex: 1;
+        margin: 0 !important;
+        height: $o-we-sidebar-content-field-colorpicker-size;
+        min-width: 0 !important;
+    }
+
+    div {
+        width: 100%;
+    }
+
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.editor.ui.scss:120]: Remove that blank line. You can also remove all blank lines between nested rules

File: addons/website/static/src/scss/secondary_variables.scss [hunk 15]
```diff
@@ -103,6 +103,9 @@ $o-has-customized-colors:
     or map-get($o-user-theme-color-palette, 'warning')
     or map-get($o-user-theme-color-palette, 'danger'));
 
+$gray-hue: o-website-value('gray-hue');
+$gray-extra-saturation: o-website-value('gray-extra-saturation');
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/secondary_variables.scss]: Add the `!default` flag

File: addons/website/static/src/js/editor/snippets.options.js (hunk 9 above)
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js]: Strange to have to do that for each gray when you update, I think this could be refactored a bit to be prettier.

File: addons/web_editor/static/src/js/common/utils.js [hunk 16]
```diff
@@ -279,7 +278,24 @@ function _colorHEXtoHSL(hex) {
...
+function _colorHSLToHEX(h, s, l) {
+    l /= 100;
+    const a = s * Math.min(l, 1 - l) / 100;
+    const f = n => {
+        const k = (n + h / 30) % 12;
+        const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
+        return Math.round(255 * color).toString(16).padStart(2, '0'); // convert to Hex and prefix "0" if needed
+    };
+    return `#${f(0)}${f(8)}${f(4)}`;
+}
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/utils.js:298]: Same: already something which is defined in colorpicker.js (from rgba, but it can be converted first).

File: addons/web_editor/static/src/scss/secondary_variables.scss [hunk 17]
```diff
@@ -48,18 +48,7 @@ $o-theme-color-palette: map-merge($o-base-theme-color-palette, o-map-omit($o-the
...
-@function adjust-gray-color($gray) {
-    @return adjust-color(change-color($gray, $hue: $gray-hue * 1deg), $saturation: $gray-extra-saturation * 1%);
-}
-$o-adjusted-gray-color-palette: ();
-@each $key, $color in $o-base-gray-color-palette {
-    $o-adjusted-gray-color-palette: map-merge($o-adjusted-gray-color-palette,
-            (base-#{$key}: $color, $key: adjust-gray-color($color))
-    );
-}
-$o-gray-color-palette: map-merge($o-adjusted-gray-color-palette, o-map-omit($o-gray-color-palette));
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/secondary_variables.scss]: OK your first commit creates this, your second removes it... it makes review complicated and history useless. Try to refactore your history so that it is clear.

File: addons/website/static/src/js/editor/snippets.options.js [hunk 18]
```diff
@@ -940,20 +949,43 @@ options.registry.OptionsTab = options.Class.extend({
...
+        return this._makeSCSSCusto(url, values);
+    },
+    /**
+     * @override
+     */
+    async _customizeWebsiteByType(widgetValue, params, type) {
+        if (type === "grays") {
+            return this._customizeWebsiteGrayParameters(widgetValue, params);
+        }
+        return this._super(...arguments);
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:988]: It's ok to have split `_customizeWebsite` to do that but feels like a bit overkill. Add the possibility to customize grays in the base Class directly, it would be easier. No need to split and everything..

File: addons/website/static/src/scss/primary_variables.scss [hunk 19]
```diff
@@ -16,10 +16,14 @@ $o-base-color-palette: map-merge($o-base-color-palette, (
...
 $o-user-color-palette: () !default;
 
-// By default, all user theme color palette values are null. Each null value
+// By default, all user theme & gray color palette values are null. Each null value
 // is automatically replaced with corresponsing color of chosen theme color
 // palette.
 $o-user-theme-color-palette: () !default;
+$o-user-gray-palette-parameters: (
+        'gray-hue': 210,
+        'gray-extra-saturation': 0,
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/primary_variables.scss:25]: Too many indentation, use the same as everywhere

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Ok I think some of my comments don't make sense because I commented commit by commit and your history was a bit messy (the second commit overrides many stuff of the first one). Maybe your commits do not make sense on their own and you should squash them? 

Review your commit message to mention the right apps (web-editor does not exist) and respect the limit for character per line).

Note you also misunderstood what I said about the commit which comes from another PR. You added "-- waiting for merge -- #66812" in a commit which belongs to this PR, what I wanted to say is to add *as commit title* (not after) and in the commit which belongs to the foreign PR.

File: addons/website/static/src/scss/secondary_variables.scss (hunk 15 above)
Comment (oussamamessaoudi) [reply addons/website/static/src/scss/secondary_variables.scss]: The value of `$gray-huy `and `$gray-extra-saturation` must be synchronized with the values chosen by the user, IMO the value of the two variables must always be replaced by the new ones (so no !default).

Comment (qsm-odoo) [reply addons/website/static/src/scss/secondary_variables.scss]: It makes sense indeed :+1:

File: addons/website/static/src/js/editor/snippets.options.js (hunk 18 above)
Comment (oussamamessaoudi) [reply addons/website/static/src/js/editor/snippets.options.js:990]: IMO, if I add gray persistence logic (`_customizeWebsite`) in the base Class, I will also need to add `_computeWidgetState` logic in the base Class as well.
so should i put all the gray logic in the base Class or just persistence logic?

https://github.com/odoo/odoo/blob/0816450bb6583f4b81b96669b7e972b9990a8826/addons/website/static/src/js/editor/snippets.options.js#L1022

Comment (qsm-odoo) [reply addons/website/static/src/js/editor/snippets.options.js:990]: Mmh, ok, let's keep the split you made :+1:

**Review by qsm-odoo — APPROVED** [review]
@oussamamessaoudi I changed a large part of the work. The `gray-hue` and `gray-extra-saturation` were still to be "normal" website values so saved in the website values palette like you did originally... not sure at which point you changed that. Then as we decided to compute the grays on the client side for the previews, we can indeed save those computed grays from the client side instead of duplicating the logic on the scss... but those grays were then to be saved in an `$o-user-gray-color-palette` like the normal and theme ones (and so simply reuse the `_customizeWebsiteColor`). Then there was some stuff to do to handle color palette reset.

Actually we now even have the possibility to simplify this even more: the gray-hue and gray-extra-saturation values saved on the server side are totally useless -> the client side can re-find them from the gray colors easily. And this would actually solved a small preview bug (however that bug is the same for status colors in 14.0 so I chose to ignore it for now). As we need this work for further merge of other PR for the freeze, I'll merge it this way and see later for those simplifications.

@robodoo r+

**Comment by xmo-odoo** [comment]
@robodoo retry

**Comment by qsm-odoo** [comment]
Improved with https://github.com/odoo/odoo/commit/e08184103c42eaf7f6bfb70655cb90b51e79913a
